mender.go

```diff
@@ -104,20 +132,61 @@ func (m mender) GetCurrentImageID() string {
 	return imageID
 }
 
-func (m *mender) updateState() error {
+func (m *mender) changeState(state MenderState) {
+	log.Infof("Mender state: %v -> %v", m.state, state)
+	m.state = state
+}
+
+func (m *mender) hasUpgrade() (bool, error) {
 	env, err := m.ReadEnv("upgrade_available")
 	if err != nil {
-		return err
+		return false, err
 	}
 	upgradeAvailable := env["upgrade_available"]
 
 	// we are after update
 	if upgradeAvailable == "1" {
-		m.state = MenderRunningWithFreshUpdate
-		return nil
+		return true, nil
 	}
-	m.state = MenderFreshInstall
-	return nil
+	return false, nil
+}
+
+func (m *mender) updateState() {
+
+	newstate := MenderStateUnknown
+	var merr error
+
+	switch m.state {
+	case MenderStateInit:
+		if m.needsBootstrap() {
+			if err := m.doBootstrap(); err != nil {
+				newstate = MenderStateError
+				merr = err
+			} else {
+				newstate = MenderStateBootstrapped
+			}
+		} else {
+			newstate = MenderStateBootstrapped
+		}
+	case MenderStateBootstrapped:
+		upg, err := m.hasUpgrade()
+		if err != nil {
+			newstate = MenderStateError
+			merr = err
```

Comment:
Yes, it will exit at this point. I think adding some sort of checking if error is transient may help here. Something in the line of:
```
func IsErrTransient(e error) bool {
....
}
```
and making sure to define sensible errors.

Anyway, we need to discuss the agent state machine as it's not that clear to me. The previous code would just enter `MenderStateUnknown` and continue working, even though the error was caused by a failure reading uboot environment (so it's not really transient). We ended up in an undefined state, cannot read uboot environment, but keep on poking the server for new updates, you see that there's something fishy around here?

Since it looks like a larger job, I propose to add the transient error check at this point, followed by opening a JIRA ticket to have the state machine defined in more detail.

@pasinskim what's your take on this?